source/Higgs/UntruncatedAtMostTwo.lagda

```diff
@@ -0,0 +1,244 @@
+        (pt : propositional-truncations-exist)
+       where
+
+open import Higgs.Rigidity fe pe
+open import Higgs.InvolutionTheorem fe pe
+open import Higgs.AutomorphismsOfOmega fe pe
+open import Higgs.AutomorphismsOfOmegaWEM fe pe pt
+
+open Conjunction
+open Implication fe
+open Universal fe
+
+\end{code}
+
+Assuming function extensionality, having an untruncated pigeonhole principle
+reflects discreteness to the codomain.
+
+The core method is to note that the pidgeonhole function must respect equality
+of functions, so we produce a pair of functions, where equality of their results
+holds if and only if the two elements of the codomain are equal.
+
+\begin{code}
+
+constantly : {𝓤 𝓥 : Universe}
+           → {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+           → Y → X → Y
+constantly const-y _ = const-y
+
+almost-constantly : {𝓤 𝓥 : Universe}
+                  → {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+                  → is-discrete X
+                  → X → Y → Y → X → Y
+almost-constantly X-discrete pt-x pt-y const-y x' =
+  cases
+   (λ _ → pt-y) (λ _ → const-y)
+   (X-discrete x' pt-x)
+
+almost-constantly-is-constant
+ : {𝓤 𝓥 : Universe}
+ → {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+ → (X-discrete : is-discrete X)
+ → (x : X)
+ → (y y' : Y)
+ → y ＝ y'
+ → constantly y ＝ almost-constantly X-discrete x y' y
+almost-constantly-is-constant {_} {_} {X} {Y} X-discrete x y _ refl = dfunext fe I
+ where
+  I : (x' : X)
+    → constantly y x' ＝ almost-constantly X-discrete x y y x'
+  -- possibly rewrite this with dep-cases?
+  -- it's more obvious and spartan but much more awkward
+  I x' with X-discrete x' x
+  ... | inl _ = refl
+  ... | inr _ = refl
+
+at-most-discrete-gives-discrete
+  : {𝓤 𝓥 : Universe}
+  → (X : 𝓤 ̇) (Y : 𝓥 ̇)
+  → is-discrete X
+  → ((f : X → Y) → f has-a-repetition)
+  → is-discrete Y
+at-most-discrete-gives-discrete X Y X-discrete f-ph y y' = V VI
+ where
+
+  repeat-indices : (X → Y)
+                 → X × X
+  repeat-indices f =
+    (λ (x , x' , _) → (x , x'))
+    (f-ph f)
+
+  repeat-is-repeat : (f : X → Y)
+                   → let (x , x') = repeat-indices f in
+                   f x ＝ f x'
+  repeat-is-repeat f =
+    let (x , x' , _ , pf) = f-ph f
+    in pf
+
+  repeat-distinct : (f : X → Y)
+                → let (x , x') = repeat-indices f in
+                x ≠ x'
+  repeat-distinct f =
+    let (x , x' , pf , _) = f-ph f
+    in pf
+
+  f₁ = constantly y
+  ix₁ = repeat-indices f₁
+  f₂ = almost-constantly X-discrete (pr₁ ix₁) y' y
+  ix₂ = repeat-indices f₂
+
+  I : y ＝ y' → ix₁ ＝ ix₂
+  I e = ap repeat-indices
+    (almost-constantly-is-constant X-discrete (pr₁ ix₁) y y' e)
+
+  II : (x : X)
+     → (pr₁ ix₁ ≠ x)
+     → y ＝ f₂ x
+  II x ne with X-discrete x (pr₁ ix₁)
+  ... | inl e = 𝟘-elim (ne (e ⁻¹))
+  ... | inr _ = refl
+
+  III : f₂ (pr₁ ix₁) ＝ y'
+  III with X-discrete (pr₁ ix₁) (pr₁ ix₁)
+  ... | inl e = refl
+  ... | inr ne = 𝟘-elim (ne refl)
+
+  IV : ix₁ ＝ ix₂ → y ＝ y'
+  IV e =
+    y            ＝⟨ II (pr₂ ix₁) (repeat-distinct f₁) ⟩
```

Comment:
Align the equality symbols and the end of equality proof symbol.

Reply:
Fixed